tuf/client_rework/updater_rework.py

```diff
@@ -35,27 +36,22 @@ class Updater:
     """
     Provides a class that can download target files securely.
 
-    Attributes:
-        metadata:
-
-        repository_name:
-
-        mirrors:
-
-        fetcher:
-
-        consistent_snapshot:
+    TODO
     """
 
     def __init__(
         self,
         repository_name: str,
-        repository_mirrors: Dict,
+        metadata_base_url: str,
+        target_base_url: Optional[str] = None,
```

Comment:
Instead of a separate base URL for metadata and targets, should we imitate the mirrors implementation we are removing which has a base URL + 'metadata' for metadata downloads and base URL + 'targets' for target file downloads? Something like:
```
# constructor parameters
repository_base_url: str,
default_metadata_dir: Optional[str] = None,
default_targets_dir: Optional[str] = None

# ensure sane defaults
if default_metadata_dir is None:
  default_metadata_dir = 'metadata'
self.metadata_base_url = parse.urljoin(repository_base_url, default_metadata_dir)
# etc
```
This gets slightly less intuitive when it comes to `download_target()`, where I think we want to provide a replacement for repository_base_url + default_targets_dir ? Still, I think having a base URL for the repository plus default paths joined to that base for targets and metadata makes sense. What do you think?

Reply:
`default_[metadata|targets]_path` may be a better name?

Reply:
There's two cases to optimize for I guess: the reference "example" implementation and code that gets used in production.

For the first "example code" case: I'd argue this is easy to understand (and communicates the possibility of using separate hosts or whatever directories as a bonus):
`Updater("/local/metadata/", "http://localhost:8080/metadata/", "http://localhost:8080/targets/")
`
note that the current PR still uses repository_name as first arg so the example above is from the future -- I'm demonstrating what the end result should be IMO: the only things the client decides are
 * local metadata directory
 * remote metadata base url
 * (optional) remote targets base url

I'm not sure adding directories to this makes it simpler (that said the example code is definitely confusing right now though: I pinky swear I'm going to improve that in coming PRs).


For the second case (production use), using a single URL and two paths makes one case more complex: if only one target host is used (so overriding the base url would not be needed in `download_target()`) but the host is not the same as metadata host. We don't have a lot of statistical data but this feels like it's not an edge case: We already know that hosting metadata of a regularly timestamping TUF instance sort of requires "actively running code on and maintaining a server" but hosting the target files is a matter of dumping them on a file server. I would expect a "two host" model to be a logical choice for production setups.

Reply:
> We already know that hosting metadata of a regularly timestamping TUF instance sort of requires "actively running code on and maintaining a server" but hosting the target files is a matter of dumping them on a file server. I would expect a "two host" model to be a logical choice for production setups.

This is a well reasoned argument, I'm convinced. Thanks for taking the time to explore it with me.